File: replaceWords-LC 648.py
# Time Complexity = O(n * l) + O(m * k), where n = no. of words in dictionaty, l = avg length of those words, m = no. of words in the sentence, k = avg length of those words

# Space Complexity = O(n * l)


class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False
    
    
class Solution:
    def __init__(self):
        self.root = TrieNode()
        
    
    def insert(self, word):
        curr = self.root
        for char in word:
            c = ord(char) - ord('a')
            if curr.children[c] == None:
                curr.children[c] = TrieNode()
                #curr.children[c].val = char         # Optional
            
            curr = curr.children[c]
        
        curr.isEnd = True
        
        
    def search(self, word):
        curr = self.root
        string = ""
        for char in word:
            c = ord(char) - ord('a')
            
            # If no trie found, return empty string, else append the char to string to give out the smaller replacement of the word
            if curr.children[c] == None :
                return ""
            
            string += char
            curr = curr.children[c]
            
            # If isEnd is True, it means that a smaller replacement is found, return the formed string
            if curr.isEnd == True:
                return string
        
        return ""
        
        
    
        
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        for word in dictionary:
            self.insert(word)
       
        stringList = sentence.split()
        final = ""
        
        for i in range(len(stringList)):
            # to separate the formed sentence with spaces
            if (i != 0):
                final += " "
                
            strs = stringList[i]
            ret = self.search(strs)
            
            if (ret != ""):
                final += ret
            else:
                final += strs
        
        return final
                
            
            
        
        
        
        

File: ReplaceWordsLC648.java
/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */


class Solution {



    //Trie Approach
    //Time Complexity: O(n*l + m*l), n = length of sentence, m = number of words in dictionary, l = length of words
    //Space Complexity: O(m*l + n*l)
    class TrieNode {

        TrieNode[] children;
        boolean end;

        public TrieNode(){
            children = new TrieNode[26];
            end = false;
        }
    }

    class Trie {

        private TrieNode root;

        public Trie() {

            root = new TrieNode();
        }

        public void insert(String word) {

            TrieNode node = root;

            for(int i=0; i<word.length(); i++){

                int index = word.charAt(i) - 'a';
                if(node.children[index] == null){
                    node.children[index] = new TrieNode();

                }
                node = node.children[index];
            }
            node.end = true;
        }

        public String replace(String word){

            TrieNode node = root;
            String temp="";

            for(int i=0; i<word.length(); i++){

                int index = word.charAt(i) - 'a';

                if(node.end == true){
                    return temp;
                }

                if(node.children[index] == null){
                    return word;
                }
                temp += word.charAt(i);
                node = node.children[index];
            }
            return word;
        }
    }

    public String replaceWords(List<String> dictionary, String sentence) {

        Trie obj = new Trie();

        for(String dict : dictionary){
            obj.insert(dict);
        }

        String[] words = sentence.split(" ");
        String result = "";
        for(String word: words){

            if(result != ""){
                result += " ";
            }
            result += obj.replace(word);

//             String temp = obj.replace(word);

//             if(temp!= null){
//                 result += temp;
//             }
//             else{
//                 result += word;
//             }
        }
        return result;
    }




    //HashSet Approach
    //Time Complexity: O(n*k*k), n = number of words in sentence, k = average length of words in sentence
    //Space Complexity: O(m), m = number of words in dictionary

//     public String replaceWords(List<String> dictionary, String sentence) {

//         Set<String> set = new HashSet<>(dictionary);

//         StringBuilder result = new StringBuilder();
//         String[] strArr = sentence.split(" ");

//         for(int i=0; i<strArr.length; i++){

//             String word = strArr[i];

//             if(i>0) result.append(" ");

//             int prevLen = result.length();

//             for(int j=0; j<word.length(); j++){

//                 String currWord = word.substring(0, j+1);

//                 if(set.contains(currWord)){
//                     result.append(currWord);
//                     break;
//                 }
//             }

//             if(prevLen == result.length()){
//                 result.append(word);
//             }
//         }
//         return result.toString();
//     }
}

public class ReplaceWordsLC648 {
}
